factom_keys/serverid.py

```diff
@@ -0,0 +1,293 @@
+import re
+import ed25519
+from bitcoin import base58
+from hashlib import sha256
+
+PREFIX_LENGTH = 3
+CHECKSUM_LENGTH = 4
+BODY_LENGTH = 35
+TOTAL_LENGTH = 39
+
+
+def generate_id_key_set():
+    """
+    Generate a set of Factom Server Identity keypairs, sk1-sk4 and id1-id4.
+    :return: a list of tuples containing a random id key pair in string format: [(sk1..., id1...), (sk2..., id2...)]
+    """
+    pairs = []
+
+    for i in range(1, 5):
+        signer, verifier = ed25519.create_keypair()
+        k = ServerIdPrivateKey(signer.to_seed())
+        p = ServerIdPublicKey(verifier.to_bytes())
+        k.key_level = i
+        p.key_level = i
+        pairs.append((k.to_string(), p.to_string()))
+
+    return pairs
+
+
+def generate_key_pair(key_level: int):
+    """
+    Generate a Factom ServerId keypair at the specified level: 1-4 for (sk1, id1) - (sk4, id4).
+
+    :param key_level: the key_level as an integer, 1-4
+    :return: a tuple containing a random id key pair: (ServerIdPrivateKey, ServerIdIPublicKey)
+    """
+    signer, verifier = ed25519.create_keypair()
+    k = ServerIdPrivateKey(signer.to_seed())
+    k.key_level = key_level
+    p = ServerIdPublicKey(verifier.to_bytes())
+    p.key_level = key_level
+    return (k, p)
+
+
+def _to_base58_string(prefixed_key: bytes):
+    """
+    Convert prefixed_key bytes into sk1/id strings with a checksum.
+
+    :param prefixed_key: the sk1 private key or id public key prefixed with the appropriate bytes
+    :return: a sk1 private key string or id public key string
+    """
+    prefix = prefixed_key[:PREFIX_LENGTH]
+    if not (
+        prefix in ServerIdPublicKey.PREFIXES or prefix in ServerIdPrivateKey.PREFIXES
+    ):
+        raise InvalidPrefixError
+    temp_hash = sha256(prefixed_key[:BODY_LENGTH]).digest()
+    checksum = sha256(temp_hash).digest()[:CHECKSUM_LENGTH]
+    return base58.encode(prefixed_key + checksum)
+
+
+class BadKeyStringError(Exception):
+    pass
+
+
+class BadKeyBytesError(Exception):
+    pass
+
+
+class InvalidKeyLevelError(Exception):
+    pass
+
+
+class InvalidPrefixError(Exception):
+    pass
+
+
+class InvalidParamsError(Exception):
+    def __init__(self):
+        Exception.__init__(self, "Only provide one of seed bytes or key_string, not both")
```

Comment:
Rather than baking this into the function, it can just be:

```python
class InvalidParamsError(ValueError):
    pass
```

and usage would be: `raise InvalidParamsError("Only provide one of seed bytes or key_string, not both")` so that we can be flexible with what the error is used for in the future. (I am likely going to move all these errors into a shared errors.py file)